ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -52,6 +55,17 @@ pub const TARGET_PROOF_SIZE: usize = 8192;
 #[cfg(not(test))]
 pub const TARGET_PROOF_SIZE: usize = 50_000_000;
 
+/// Maximum proof recursion depth.
+//
+// This is a hard limit. Each GF(2) multiply generates four G values and four H values,
+// so the restriction is:
+// $$ large_recursion_factor * small_recursion_factor ^ (depth - 1) >= 4 * target_proof_size $$
+// With large_recursion_factor = 32 and small_recursion_factor = 8, this means:
+// $$ depth >= log_8 target_proof_size $$
+// Because multiplication intermediate storage gets rounded up to blocks of 256, leaving
+// some margin is advised.
+pub const MAX_PROOF_RECURSION: usize = 9;
```

Comment:
The math written here is not quite right. With the correct math, recursion depth of 9 corresponds to TARGET_PROOF_SIZE of about 117M. Without quicksort batching, this is not quite enough. So either we can wait for quicksort batching, or we can temporarily increase MAX_PROOF_RECURSION slightly. I lean towards the former unless we identify an urgency to take this change.